drivers/serial/uart_native_pty.c

```diff
@@ -209,6 +247,163 @@ static int np_uart_poll_in(const struct device *dev, unsigned char *p_char)
 	}
 }
 
+#ifdef CONFIG_UART_ASYNC_API
+
+static int np_uart_callback_set(const struct device *dev, uart_callback_t callback, void *user_data)
+{
+	struct native_pty_status *data = dev->data;
+
+	data->async.user_callback = callback;
+	data->async.user_data = user_data;
+
+	return 0;
+}
+
+static void np_uart_tx_done_work(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct native_pty_status *data =
+		CONTAINER_OF(dwork, struct native_pty_status, async.tx_done);
+	struct uart_event evt;
+	unsigned int key = irq_lock();
+
+	evt.type = UART_TX_DONE;
+	evt.data.tx.buf = data->async.tx_buf;
+	evt.data.tx.len = data->async.tx_len;
+
+	(void)nsi_host_write(data->out_fd, evt.data.tx.buf, evt.data.tx.len);
+
+	data->async.tx_buf = NULL;
+
+	if (data->async.user_callback) {
+		data->async.user_callback(data->async.dev, &evt, data->async.user_data);
```

Comment:
Maybe not critical for native but  normally user callback should not be called with interrupts locked.

Reply:
The goal here was to prevent the system workqueue item being preempted by some other thread while this function was running, since in a "normal" driver this logic would be running from an interrupt context. Happy to use a different solution if you have one.